doc/architecture.md

```diff
@@ -0,0 +1,206 @@
+      same filename as the asset, while a direct, unsigned S3 URL will be
+      downloaded as a file with the same filename as the S3 key, which is
+      usually the blob ID and not user-friendly.  Hence, by default, a `GET`
+      request to `dandidav` for a blob asset will be responded to with a
+      redirect to the Archive API download URL.  Unfortunately, certain WebDAV
+      clients (i.e., [davfs2](https://savannah.nongnu.org/bugs/?65376)) do not
+      support WebDAV servers that return redirects to other redirects, so the
+      `--prefer-s3-redirects` CLI option was added to `dandidav` to instead
+      make these `GET` requests redirect directly to the unsigned S3 URLs.
+
+        - Note that HTML listings of a collection's children will always link
+          blob assets to their Archive API download URLs, regardless of
+          `--prefer-s3-redirects`, as these listings are only returned for
+          `GET` requests to a collection, which WebDAV clients don't do.
+          (These listings link blob assets and Zarr entries directly to
+          download URLs rather than `dandidav` resource URLs because doing the
+          latter would result in subsequent requests to `dandidav` that just
+          duplicate work.)
+
+        - Note that this does not apply to download URLs for Zarr entries, as
+          their S3 URLs are assumed to always have the same filename as the
+          entry, and thus signing to add a `Content-Disposition` is
+          unnecessary.
+
+- [`DandiDav::get_resource()`][] and
+  [`DandiDav::get_resource_with_children()`][] perform method calls on a
+  `DandiClient` or `ZarrManClient` instance (as appropriate) belonging to the
+  `DandiDav` instance in order to fetch information about the specified
+  resource(s).
+
+
+`DandiClient`
+-------------
+
+- The [`DandiClient`][] type is used to retrieve information about resources on
+  an Archive instance (the one specified via the `--api-url` option on program
+  invocation) via a combination of the Archive's API and (for entries & folders
+  inside Zarrs) requests to an S3 bucket.  It is the data source for the
+  `/dandisets/` hierarchy served by `dandidav`.
+
+    - Information about child resources of Zarrs needs to be fetched via S3
+      directly rather than using the Archive API because the latter currently
+      does not provide an efficient way to list folders within a Zarr.
+
+- The `DandiClient` API is modeled somewhat after
+  [`octocrab`](https://github.com/XAMPPRocky/octocrab): To operate on an API
+  resource, an endpoint object is requested from the `DandiClient` (possibly
+  indirectly via a super-endpoint), and the endpoint's methods are called to
+  obtain one or more objects that provide information about the resource or its
+  children, but these resource objects are distinct from the endpoint objects
+  and do not have any methods for making API calls themselves.
```

Comment:
any caching here? would navigating dandiset folder could requesting listing for a folder or entire dandiset?

Reply:
No caching, but see https://github.com/dandi/dandidav/issues/6.  Viewing a folder inside a Dandiset results in a request to `/dandisets/{dandiset_id}/versions/{version_id}/assets/paths/?path_prefix={folder}` to list just the contents of that folder.